ui/ducks/send/send.js

```diff
@@ -513,27 +533,145 @@ export const computeEstimatedGasLimit = createAsyncThunk(
 );
 
 /**
- * This method is used to keep the original logic from the gas.duck.js file
- * after receiving a gasPrice from eth_gasPrice. First, the returned gasPrice
- * was converted to GWEI, then it was converted to a Number, then in the send
- * duck (here) we would use getGasPriceInHexWei to get back to hexWei. Now that
- * we receive a GWEI estimate from the controller, we still need to do this
- * weird conversion to get the proper rounding.
+ * @typedef {Object} Asset
+ * @property {AssetTypesString} type - The type of asset that the user
+ *  is attempting to send. Defaults to 'NATIVE' which represents the native
+ *  asset of the chain. Can also be 'TOKEN' or 'COLLECTIBLE'.
+ * @property {string} balance - A hex string representing the balance
+ *  that the user holds of the asset that they are attempting to send.
+ * @property {TokenDetails} [details] - An object that describes the
+ *  selected asset in the case that the user is sending a token or collectibe.
+ *  Will be null when asset.type is 'NATIVE'.
+ * @property {string} [error] - Error to display when there is an issue
+ *  with the asset.
+ */
+
+/**
+ * Takes in a partial asset that must have at least a type specified, and will
+ * attempt to get all the other details necessary to create a fully qualifed
+ * asset object. The bare minimum amount of data known is either type and the
+ * contract address specified via the asset.details.address property, or the
+ * type and the transaction's hex encoded data property which can be decoded
+ * to ascertain the address. With those pieces all other asset.details fields
+ * can be populated.
  *
- * @param {string} gasPriceEstimate
- * @returns {string}
+ * @param {Pick<Asset, 'type' | 'details'>} asset - min amount of asset fields.
+ * @param {string} address - The address to get the balance of for the asset.
+ * @param {string} nativeBalance - The native asset balance for the address.
+ * @param {object[]} tokens - The tokens the user has in their wallet.
+ * @param {string} [transactionData] - Hex encoded transaction data.
+ * @returns {(dispatch: import('redux').Dispatch) => Asset}
  */
-function getRoundedGasPrice(gasPriceEstimate) {
-  const gasPriceInDecGwei = conversionUtil(gasPriceEstimate, {
-    numberOfDecimals: 9,
-    toDenomination: GWEI,
-    fromNumericBase: 'dec',
-    toNumericBase: 'dec',
-    fromCurrency: ETH,
-    fromDenomination: GWEI,
-  });
-  const gasPriceAsNumber = Number(gasPriceInDecGwei);
-  return getGasPriceInHexWei(gasPriceAsNumber);
+function getAssetDetailsAndBalance(
```

Comment:
[self review] I would love for this function to be non side effect producing (currently it opens modals, shows loading indication, etc)... The only way I thought to do that was to interpret the error message in the consumer... but then not only are we using errors for control flow, we're also requiring duplication of that flow in multiple places (every callsite). @adonesky1 if you get a minute could you tell me if the logic in here matches the old upsetSendAsset flow?  You did most of the work on collectibles which is where i'm the least confident.